core/src/main/scala/org/apache/spark/util/ListenerBus.scala

```diff
@@ -17,53 +17,222 @@
+  Unit = synchronized {
+    var listenerEventExecutor = eventGroupToEventExecutors.get(eventListenerGroup)
+    if (listenerEventExecutor == null) {
+      listenerEventExecutor =
+        new ListenerEventExecutor[L](listener.getClass.getName, eventQueueSize)
+      eventGroupToEventExecutors.put(eventListenerGroup, listenerEventExecutor)
+
+    }
+    listenerEventExecutor.addListener(listener)
+    if (started.get()) {
+      listenerEventExecutor.start
+    }
   }
 
   /**
    * Remove a listener and it won't receive any events. This method is thread-safe and can be called
    * in any thread.
    */
   final def removeListener(listener: L): Unit = {
-    listeners.remove(listener)
+    for (eventExecutor <- eventGroupToEventExecutors.values().asScala) {
+      eventExecutor.removeListener(listener)
+    }
+  }
+
+   /**
+    * For testing only. Returns whether there is any event pending to be processed by
+    * any of the existing listener
+    */
+  def isListenerBusEmpty: Boolean = {
+    for (eventExecutor <- eventGroupToEventExecutors.values().asScala) {
+      if (!eventExecutor.isEmpty) {
+        return false
+      }
+    }
+    true
   }
 
   /**
-   * Post the event to all registered listeners. The `postToAll` caller should guarantee calling
-   * `postToAll` in the same thread for all events.
+   * Posts the event to all registered listeners. This is an async call and it does not
+   * processes the event itself. Processing of the event is done in a separate thread in
+   * the {@link ListenerEventExecutor}.
    */
   final def postToAll(event: E): Unit = {
-    // JavaConverters can create a JIterableWrapper if we use asScala.
-    // However, this method will be called frequently. To avoid the wrapper cost, here we use
-    // Java Iterator directly.
-    val iter = listeners.iterator
-    while (iter.hasNext) {
-      val listener = iter.next()
-      try {
-        doPostEvent(listener, event)
-      } catch {
-        case NonFatal(e) =>
-          logError(s"Listener ${Utils.getFormattedClassName(listener)} threw an exception", e)
+    for (listenerEventProcessor <- eventGroupToEventExecutors.values().asScala) {
+      // JavaConverters can create a JIterableWrapper if we use asScala.
+      // However, this method will be called frequently. To avoid the wrapper cost, here we use
+      // Java Iterator directly.
+      val iter = listenerEventProcessor.listeners.iterator()
+      while (iter.hasNext) {
+        val listener = iter.next()
+        listenerEventProcessor.submit(new Runnable {
+          override def run(): Unit = Utils.tryLogNonFatalError {
+            doPostEvent(listener, event)
+          }
+        })
+      }
+    }
+  }
+
+  /**
+   * For testing only. Post the event to all registered listeners.
+   * This guarantees processing the event in the same thread for all
+   * events.
+   */
+  final def postToAllSync(event: E): Unit = {
```

Comment:
Similarly to the above, this feels hacky. If you follow my previous suggestion, you could return a `Future` from the `submit()` method, and this method would just become a call like `submit(...).get()`.